### ISSUE TITLE
Deepspeed custom lr scheduler load resets

### ISSUE THREAD

**Issue by KeremTurgutlu** [huggingface/accelerate#1673]
### System Info

```Shell
accelerate==0.20.3
numpy==1.22.2
torch==1.14.0a0+44dac51
deepspeed==0.9.5

[2023-07-01 01:16:19,936] [INFO] [real_accelerator.py:110:get_accelerator] Setting ds_accelerator to cuda (auto detect)

Copy-and-paste the text below in your GitHub issue

- `Accelerate` version: 0.20.3
- Platform: Linux-5.4.0-147-generic-x86_64-with-glibc2.29
- Python version: 3.8.10
- Numpy version: 1.22.2
- PyTorch version (GPU?): 1.14.0a0+44dac51 (True)
- PyTorch XPU available: False
- System RAM: 354.07 GB
- GPU type: NVIDIA A100-SXM4-80GB
- `Accelerate` default config:
        - compute_environment: LOCAL_MACHINE
        - distributed_type: DEEPSPEED
        - use_cpu: False
        - num_processes: 4
        - machine_rank: 0
        - num_machines: 1
        - rdzv_backend: static
        - same_network: True
        - main_training_function: main
        - deepspeed_config: {'deepspeed_config_file': '/workspace/workdir/zero_stage0_minimal.json', 'zero3_init_flag': True}
        - downcast_bf16: no
        - tpu_use_cluster: False
        - tpu_use_sudo: False
        - tpu_env: []

Linux fd36dc77cdff 5.4.0-147-generic #164-Ubuntu SMP Tue Mar 21 14:23:17 UTC 2023 x86_64 x86_64 x86_64 GNU/LinuxPython 3.8.10 (default, Nov 14 2022, 12:59:47) 
```

```json
{
    "bf16": {
        "enabled": true
    },
    "zero_optimization": {
        "stage": 0,
        "offload_optimizer": {
            "device": "none",
            "pin_memory": true
        },
        "offload_param": {
            "device": "none",
            "pin_memory": true
        },
        "allgather_partitions": true,
        "allgather_bucket_size": 2e8,
        "overlap_comm": true,
        "reduce_scatter": true,
        "reduce_bucket_size": 5e8,
        "contiguous_gradients": true
    },
    "gradient_accumulation_steps": 4,
    "gradient_clipping": 1.0,
    "steps_per_print": 1000,
    "train_batch_size": "auto",
    "train_micro_batch_size_per_gpu": "auto",
    "wall_clock_breakdown": false
}
```

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [X] One of the scripts in the examples/ folder of Accelerate or an officially supported `no_trainer` script in the `examples` folder of the `transformers` repo (such as `run_no_trainer_glue.py`)
- [ ] My own task or dataset (give details below)

### Reproduction

Run the example [example script](https://github.com/huggingface/accelerate/blob/main/examples/by_feature/deepspeed_with_config_support.py) with `cosine` scheduler save after some steps and resume.

#### Issue

I am using the following functions to save and resume deepspeed training similar to the original [example script](https://github.com/huggingface/accelerate/blob/main/examples/by_feature/deepspeed_with_config_support.py).

```python
# New Code #
def checkpoint_model(checkpoint_folder, ckpt_id, model, epoch, last_global_step, **kwargs):
    """Utility function for checkpointing model + optimizer dictionaries
    The main purpose for this is to be able to resume training from that instant again

    NOTE: Important: all processes must call this method and not just the process with rank 0.
    It is because each process needs to save its master weights and scheduler+optimizer states. 
    This method will hang waiting to synchronize with other processes if it’s called just for the
    process with rank 0.
    """
    checkpoint_state_dict = {
        "epoch": epoch,
        "last_global_step": last_global_step,
    }
    # Add extra kwargs too
    checkpoint_state_dict.update(kwargs)
    success = model.save_checkpoint(checkpoint_folder, ckpt_id, checkpoint_state_dict)
    status_msg = f"checkpointing: checkpoint_folder={checkpoint_folder}, ckpt_id={ckpt_id}"
    if success:
        logger.info(f"Success {status_msg}")
<img width="1408" alt="Screenshot 2023-06-30 at 6 14 21 PM" src="https://github.com/huggingface/accelerate/assets/19826777/c0e35a33-d4b9-4d22-8313-71d98d919ea8">

    else:
        logger.warning(f"Failure {status_msg}")
    return


# New Code #
def load_training_checkpoint(model, load_dir, tag=None, **kwargs):
    """Utility function for checkpointing model + optimizer dictionaries
    The main purpose for this is to be able to resume training from that instant again

    NOTE: Important: under ZeRO3, one cannot load checkpoint with engine.load_checkpoint()
    right after engine.save_checkpoint(). It is because engine.module is partitioned, 
    and load_checkpoint() wants a pristine model. If insisting to do so, please
    reinitialize engine before load_checkpoint(). 
    
    Assuming resuming training invovles reinitializing engine, so it is fine to do so.
    """
    _, checkpoint_state_dict = model.load_checkpoint(load_dir, tag=tag, **kwargs)
    logger.info(f"Checkpoint state dict keys: {checkpoint_state_dict.keys()}")
    epoch = checkpoint_state_dict["epoch"]
    last_global_step = checkpoint_state_dict["last_global_step"]
    del checkpoint_state_dict
    return (epoch, last_global_step)
```

I am using HF `get_scheduler`, however the scheduler doesn't seem to be resuming from where it was at, e.g. I am using cosine annealing.

Logging: `"train/learning_rate": lr_scheduler.get_last_lr()[0]`

<img width="1408" alt="Screenshot 2023-06-30 at 6 14 21 PM" src="https://github.com/huggingface/accelerate/assets/19826777/98a026e0-65ad-4e28-bf18-1f3e91952943">

Scheduler defined in the script:

```python
    # Creates Dummy Scheduler if `scheduler` was specified in the config file else creates `args.lr_scheduler_type` Scheduler
    if (
        accelerator.state.deepspeed_plugin is None
        or "scheduler" not in accelerator.state.deepspeed_plugin.deepspeed_config
    ):
        # NOTE: AcceleratedScheduler takes num_processes many steps per training step, when split_batches=False.
        # So we multiply the max_train_steps by accelerator.num_processes.
        # https://github.com/huggingface/accelerate/blob/e60a4243988a636bda8a6bf99044fb313d5a9e0e/src/accelerate/scheduler.py#L74
        # FIXME: lr scheduler state is not loaded correctly, seems to restart. 
        lr_scheduler = get_scheduler(
            name=args.lr_scheduler_type,
            optimizer=optimizer,
            num_warmup_steps=args.num_warmup_steps,
            num_training_steps=int(args.max_train_steps/0.8)*accelerator.num_processes, # to make min_lr = lr * 0.1.
        )
    else:
        lr_scheduler = DummyScheduler(
            optimizer, total_num_steps=int(args.max_train_steps/0.8), warmup_num_steps=args.num_warmup_steps
        )
```

### Expected behavior

LR should resume from the same training pct.

**Comment by KeremTurgutlu** [comment]
Maybe for HF schedulers this can be added:

```python
# Load LRScheduler
if isinstance(lr_scheduler, accelerate.scheduler.AcceleratedScheduler):
    lr_scheduler.scheduler.load_state_dict(os.path.join(resume_from_checkpoint, "scheduler.bin"))
```

to somewhere [here](https://github.com/huggingface/accelerate/blob/main/examples/by_feature/deepspeed_with_config_support.py#L615-L626)

**Comment by pacman100** [comment]
Hello @KeremTurgutlu, the example is a bit outdated. You can directly use `accelerator.save_state` and `accelerator.load_state` now with DeepSpeed. That should resolve the issue that you are facing and should be lot easier to use.

**Comment by pacman100** [comment]
See this test script that makes sure checkpointing is working as expected: https://github.com/huggingface/accelerate/blob/main/src/accelerate/test_utils/scripts/external_deps/test_checkpointing.py

**Comment by pacman100** [comment]
Checkout how the above script is used for DeepSpeed here: https://github.com/huggingface/accelerate/blob/main/tests/deepspeed/test_deepspeed.py#L771

**Comment by KeremTurgutlu** [comment]
Thanks for sharing these pointers! I was actually using `save_state` but not `load_state`, it looks like it is handling the AcceleratedScheduler restoring.

You mentioned that scripts are outdated but other scripts use the `load_state` method, so it wasn't handling deepspeed before and now is?

It might be a good idea to update the scripts as they are the only deepspeed examples mentioned across the documentation, or maybe put a note for other people who might run into the same issue if scripts won't be updated in near future.

Also, what exactly the test you shared asserts? I see that it runs a command but not sure what's been tested actually.

**Comment by pacman100** [comment]
> Also, what exactly the test you shared asserts? I see that it runs a command but not sure what's been tested actually.

![Screenshot 2023-07-04 at 3 22 06 PM](https://github.com/huggingface/accelerate/assets/13534540/f356f262-db41-445e-a444-93124ccf2f1f)

> It might be a good idea to update the scripts as they are the only deepspeed examples mentioned across the documentation, or maybe put a note for other people who might run into the same issue if scripts won't be updated in near future.

We welcome contributions from the community to make our docs and examples stay updated. This is a good first PR/easy PR wherein you too can contribute. Will make this as a `Good First PR` issue

**Comment by KeremTurgutlu** [comment]
Will try!
